src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -41,15 +41,108 @@ private GCHandle(object? value, GCHandleType type)
+        /// <remarks>Every handle in the cache is a GCHandleType.Pinned handle with a null target.</remarks>
+        [ThreadStatic]
+        private static FixedCapacityIntPtrStack? s_pooledPinningHandles;
+
+        /// <summary>Allocates a pinning handle whose target is the specified object.</summary>
+        /// <param name="value">The object to which the allocated handle should reference.</param>
+        /// <returns>The allocated handle.</returns>
+        private static IntPtr AllocPinnedWithPooling(object? value)
+        {
+            // If we have a cache and we can take a handle from it, do so.
+            if (s_pooledPinningHandles is FixedCapacityIntPtrStack handles && handles.TryPop(out IntPtr handle))
+            {
+                // We successfully got a handle from the cache.  Update it to point to the new target.
+                InternalSet(GetHandleValue(handle), value);
+            }
+            else
+            {
+                // We could not get a handle from the cache.  Allocate a new one.
+                handle = MarkPinned(InternalAlloc(value, GCHandleType.Pinned));
+            }
+
+            Debug.Assert(IsPinned(handle));
+            return handle;
+        }
+
+        /// <summary>"Frees" a pinning handle, which might mean actually freeing it but might also mean storing it into a cache.</summary>
+        /// <param name="handle">The handle.</param>
+        private static void FreePinnedWithPooling(IntPtr handle)
+        {
+            Debug.Assert(IsPinned(handle));
+
+            // Try to store the handle into the cache rather than actually freeing it.
+            FixedCapacityIntPtrStack handles = s_pooledPinningHandles ??= new();
+            if (handles.TryPush(handle))
+            {
+                // Clear the handle's target so that it doesn't keep the object pinned.
+                // If this were a cache accessible to multiple threads, this clearing would
+                // need to be done before calling TryAdd, as if TryAdd returns true, another
+                // thread could take ownership of the handle the moment it hits the cache.
+                // But, as the cache is only accessible to this thread, we can do it here
+                // after the TryAdd and thus avoid the additional overhead in the case where
+                // the cache is full and we have to free.
+                InternalSet(GetHandleValue(handle), null);
+            }
+            else
+            {
+                // The cache was full: actually free the handle.
+                InternalFree(GetHandleValue(handle));
+            }
+        }
 
-            if (type == GCHandleType.Pinned)
+        /// <summary>Provides a simple stack of IntPtrs with a fixed capacity.</summary>
+        /// <param name="capacity">The maximum number of items the stack can hold.</param>
+        private sealed class FixedCapacityIntPtrStack(int capacity = 8) // 8 == arbitrary limit that can be tuned over time
```

Comment:
Does this need a finalizer to prevent the handles from leaking when threads come and go?

Reply:
Yup.